aws/middleware/middleware.go

```diff
@@ -0,0 +1,101 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/internal/sdk"
+	"github.com/awslabs/smithy-go/middleware"
+	smithyHTTP "github.com/awslabs/smithy-go/transport/http"
+)
+
+// RequestInvocationIDMiddleware is a Smithy BuildMiddleware that will generate a unique ID for logical API operation
+// invocation.
+type RequestInvocationIDMiddleware struct{}
+
+// ID the identifier for the RequestInvocationIDMiddleware
+func (r RequestInvocationIDMiddleware) ID() string {
+	return "Request Invocation ID Middleware"
+}
+
+// HandleBuild attaches a unique operation invocation id for the operation to the request
+func (r RequestInvocationIDMiddleware) HandleBuild(ctx context.Context, in middleware.BuildInput, next middleware.BuildHandler) (out middleware.BuildOutput, metadata middleware.Metadata, err error) {
+	const invocationIDHeader = "amz-sdk-invocation-id"
+
+	invocationID, err := sdk.UUIDVersion4()
+	if err != nil {
+		return out, middleware.NewMetadata(), err
+	}
+
+	switch req := in.Request.(type) {
+	case *smithyHTTP.Request:
+		req.Header.Set(invocationIDHeader, invocationID)
+	default:
+		return middleware.BuildOutput{}, middleware.NewMetadata(), fmt.Errorf("unknown transport type %T", req)
+	}
+
+	return next.HandleBuild(ctx, in)
+}
+
+// AttemptClockSkewMiddleware calculates the clock skew of the SDK client
+// TODO: Could be a better name, since this calculates more then skew
+type AttemptClockSkewMiddleware struct{}
+
+func (a AttemptClockSkewMiddleware) ID() string {
+	return "Attempt Clock Skew Middlware"
+}
+
+// HandleDeserialize calculates response metadata and clock skew
+func (a AttemptClockSkewMiddleware) HandleDeserialize(ctx context.Context, in middleware.DeserializeInput, next middleware.DeserializeHandler) (out middleware.DeserializeOutput, metadata middleware.Metadata, err error) {
+	respMeta := ResponseMetadata{}
+
```

Comment:
what do you think about putting return parameters on another line similar to the smithy middleware pkg?